### PR TITLE
logalloc reclaim_timer improvements

### PR DESCRIPTION
**Pull request by cmm**
* round up reported time to microseconds
* add backtrace if stall detected
* add call site name (hierarchical when timers are nested)
* put timers in more places
* reduce possible logspam in nested timers by making sure to report on things only once and to not report on durations smaller than those already reported on

### CODE REVIEW

**Comment by avikivity** [comment]
No description.

**Comment by cmm** [comment]
added the missing description

File: utils/logalloc.cc [hunk 1]
```diff
@@ -2193,9 +2193,11 @@ class reclaim_timer {
         auto time_level = _stall_detected ? log_level::warn : log_level::debug;
         auto info_level = _stall_detected ? log_level::info : log_level::debug;
         auto MiB = 1024*1024;
+        auto msg_extra = _stall_detected ? fmt::format(", at {}", current_backtrace()) : "";
```
Comment (avikivity) [inline utils/logalloc.cc]: Generating a backtrace also takes time, so we're making the problem bigger. Oh well.

File: utils/logalloc.cc [hunk 2]
```diff
@@ -2129,7 +2129,7 @@ class reclaim_timer {
     const is_preemptible _preemptible;
     const size_t _memory_to_release;
     const size_t _segments_to_release;
-    tracker::impl& _tracker;
+    tracker::impl* _tracker;
```
Comment (avikivity) [inline utils/logalloc.cc]: Why are we making it optional?

File: utils/logalloc.cc [hunk 3]
```diff
@@ -1554,6 +1556,8 @@ class region_impl final : public basic_region_impl {
 
     void compact_segment_locked(segment* seg, segment_descriptor& desc) noexcept {
         auto seg_occupancy = desc.occupancy();
+        reclaim_timer timing_guard("compact_segment_locked", is_preemptible::no, seg_occupancy.used_space(), 1);
+
```
Comment (avikivity) [inline utils/logalloc.cc:1674]: Not sure about this. Yes, we see a stall for a specific segment compaction. But why is it interesting in which function the stall happened?

File: utils/logalloc.cc [hunk 4]
```diff
@@ -2184,6 +2185,10 @@ static void reclaim_from_evictable(region::impl& r, size_t target_mem_in_use, is
     auto deficit = shard_segment_pool.total_memory_in_use() - target_mem_in_use;
     auto used = r.occupancy().used_space();
     auto used_target = used - std::min(used, deficit + segment::size);
+    reclaim_timer timing_guard("reclaim_from_evictable", preempt, used - used_target, 0);
+    auto update_timing_guard = defer([&] {
+        timing_guard.set_memory_released(used - r.occupancy().used_space());
+    });
```
Comment (avikivity) [inline utils/logalloc.cc:2304]: Again, how will we use this?

File: utils/logalloc.cc [hunk 5]
```diff
@@ -1271,14 +1263,26 @@ reclaim_timer::reclaim_timer(const char* name, is_preemptible preemptible, size_
     , _segments_to_release(segments_to_release)
     , _tracker(tracker)
     , _debug_enabled(timing_logger.is_enabled(logging::log_level::debug))
+    , _log_region_occupancy(_debug_enabled && _tracker)
```
Comment (avikivity) [inline utils/logalloc.cc:1268]: What's the problem the refactoring is trying to solve?

File: utils/logalloc.cc [hunk 6]
```diff
@@ -2193,7 +2193,7 @@ class reclaim_timer {
         auto MiB = 1024*1024;
 
         timing_logger.log(time_level, "Reclamation cycle took {} us, trying to release {:.3f} MiB {}preemptibly",
-                          _duration / 1us, (float)_memory_to_release / MiB, _preemptible ? "" : "non-");
+                          (_duration + 500ns) / 1us, (float)_memory_to_release / MiB, _preemptible ? "" : "non-");
```
[the remark this answers is not in the document]
Comment (cmm) [reply utils/logalloc.cc]: I for one don't even remember why the time isn't just shown in microseconds, seeing as that's the effective resolution.  @bhalevy?

File: utils/logalloc.cc (hunk 2 above)
Comment (cmm) [reply utils/logalloc.cc]: @bhalevy ?

File: utils/logalloc.cc (hunk 1 above)
Comment (bhalevy) [reply utils/logalloc.cc]: Hopefully fixing the issue will provide big enough payback for the investment.
Maybe we can generate and print the backtrace only is the stall is longer than the previous maximum?

File: utils/logalloc.cc (hunk 3 above)
Comment (bhalevy) [reply utils/logalloc.cc:1674]: https://github.com/scylladb/scylla/issues/6189#issuecomment-1114678869 ?

File: utils/logalloc.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (cmm) [reply utils/logalloc.cc]: yes :)

[the remark this answers is not in the document]
Comment (bhalevy) [reply utils/logalloc.cc]: Just a convention.  Tick may also be shorter in the future.

File: utils/logalloc.cc (hunk 1 above)
Comment (avikivity) [reply utils/logalloc.cc]: yes

File: utils/logalloc.cc (hunk 3 above)
Comment (avikivity) [reply utils/logalloc.cc:1674]: We could report the reserve size, instead of making everything more complicated.

**Comment by bhalevy** [comment]
@cmm ping, please respond to Avi's comments

File: utils/logalloc.cc (hunk 5 above)
Comment (cmm) [reply utils/logalloc.cc:1268]: not sure how this commit made it out of my work branch, sorry about that.

**Comment by cmm** [comment]
v2:
* cleanups
* duration backoff within a timer hierarchy
* report reserve settings

File: utils/logalloc.cc (hunk 3 above)
Comment (avikivity) [reply utils/logalloc.cc:1685]: ...?

The added complication is not small, and it will have some small performance cost.

Comment (avikivity) [reply utils/logalloc.cc:1685]: Another way to look at it is that individual segments are anonymous, it doesn't help to know that a particular one took a long time.

File: utils/logalloc.cc (hunk 1 above)
Comment (cmm) [reply utils/logalloc.cc]: I'll have the timeout back-off within a `reclaim_timer` hierarchy.  not sure if making it back-off globally is a good idea

**Comment by cmm** [comment]
v3:
- no timing on individual segment level any more
- add more context-specific info to timeout reports
- time subsections of `compact_and_evict_locked()`

File: utils/logalloc.cc (hunk 2 above)
Comment (avikivity) [reply utils/logalloc.cc]: Well, I don't understand subordinate timers either.

**Comment by avikivity** [comment]
> v3:
> 
> * no timing on individual segment level any more
> * add more context-specific info to timeout reports
> * time subsections of `compact_and_evict_locked()`

I still see nested timing guards. Can you explain where they are and why they're important?

**Comment by cmm** [comment]
> I still see nested timing guards. Can you explain where they are and why they're important?

The one in `segment_pool::reclaim_segments()` can end up nested under the one in `tracker::impl::reclaim()`.
`tracker::impl::compact_and_evict_locked()` has an overall timing guard and 2 ones under it (for the "compact()" and "evict()" parts, respectively)

That above answers the "is" part (where) but not the "ought" part (why they're important) -- I don't understand the domain and its failure modes deeply enough to answer that confidently.  So if you tell me that none of the nested scenarios is ever interesting and/or just the top-level timers would be enough, I'll totally buy it.

**Comment by bhalevy** [comment]
ping @avikivity

**Comment by avikivity** [comment]
What I'm trying to avoid is information (in logs) and complexity (in code) overload.

If we have foo1(); foo2(); then individual guards are fine; generating their sum is not important (and it's not important if each one individually falls under the threshold and their some above - the threshold is arbitrary).

If we have while (cond) { foo1(); foo2(); }, then it becomes important, if we want to understand whether foo1() or foo2() was the problem. But I don't know that we do, and didn't check if there's such a while loop.

We want to know how many times thing X was done, so that if we divide it by the time taken we can have X's throughput. In that the nested guards don't help, because they tell you the latency of the last operation, not the total time spent doing Xs.

**Comment by cmm** [comment]
v4:
- simplify by removing the nesting logic.  in cases where reclaim timers end up nested (as can happen with the one in `reclaim_segments()`), inner times do nothing at all

**Comment by bhalevy** [comment]
@avikivity can you please review?

**Pull request merged**